File: demo-services/spring-demo/src/main/java/com/example/demo/OrdersController.java
package com.example.demo;

import com.example.authz.context.RequestContext;
import com.example.authz.web.AuthorizationFilter;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.web.client.RestTemplateBuilder;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.client.RestTemplate;

import java.util.Map;

/** Business routes — no authorization annotations; enforcement is global. */
@RestController
public class OrdersController {

    private final RestTemplate restTemplate;
    private final String downstreamUrl;

    public OrdersController(RestTemplateBuilder builder,
                            @Value("${DOWNSTREAM_URL:http://localhost:5001}") String downstreamUrl) {
        // RestTemplateBuilder applies the auto-configured OutboundPropagationInterceptor.
        this.restTemplate = builder.build();
        this.downstreamUrl = downstreamUrl;
    }

    @GetMapping("/orders/{id}")
    public Map<String, Object> getOrder(@PathVariable String id, HttpServletRequest req) {
        RequestContext ctx = (RequestContext) req.getAttribute(AuthorizationFilter.CONTEXT_ATTR);
        return Map.of("id", id, "by", ctx != null && ctx.userId() != null ? ctx.userId() : "unknown");
    }

    @GetMapping("/orders/{id}/audit")
    public Map<String, Object> auditOrder(@PathVariable String id) {
        return Map.of("id", id, "audit", true);
    }

    @PostMapping("/orders")
    public Map<String, Object> createOrder() {
        return Map.of("created", true);
    }

    @PostMapping("/orders/{id}")
    public Map<String, Object> updateOrder(@PathVariable String id, HttpServletRequest req) {
        RequestContext ctx = (RequestContext) req.getAttribute(AuthorizationFilter.CONTEXT_ATTR);
        // Echo what this downstream service saw, so callers can verify propagation.
        return Map.of(
                "id", id,
                "updated", true,
                "seenServiceName", ctx != null && ctx.serviceName() != null ? ctx.serviceName() : "none",
                "seenAuthType", ctx != null ? ctx.authenticationType().name() : "none",
                "seenCorrelationId", ctx != null ? ctx.correlationId() : "none",
                "seenUserId", ctx != null && ctx.userId() != null ? ctx.userId() : "none");
    }

    @PostMapping("/internal/reconcile")
    public Map<String, Object> reconcile(HttpServletRequest req) {
        RequestContext ctx = (RequestContext) req.getAttribute(AuthorizationFilter.CONTEXT_ATTR);
        return Map.of("reconciled", true, "by", ctx != null ? ctx.serviceName() : "unknown");
    }

    @PostMapping("/orders/{id}/forward")
    public Map<String, Object> forwardOrder(@PathVariable String id, HttpServletRequest req) {
        // The RestTemplate has the OutboundPropagationInterceptor auto-configured,
        // so it attaches the user JWT, service token, and correlation/request ids.
        try {
            var response = restTemplate.postForEntity(
                    downstreamUrl + "/orders/" + id,
                    null,
                    Map.class);
            return Map.of(
                    "forwarded", true,
                    "downstreamStatus", response.getStatusCode().value(),
                    "downstream", response.getBody());
        } catch (org.springframework.web.client.HttpClientErrorException e) {
            return Map.of(
                    "forwarded", false,
                    "downstreamStatus", e.getStatusCode().value(),
                    "downstream", e.getResponseBodyAs(Map.class));
        } catch (Exception e) {
            return Map.of(
                    "forwarded", false,
                    "error", "downstream service unavailable");
        }
    }
}

File: tests/demo-services/nestjs-demo/src/orders.controller.ts
import { Body, Controller, Get, Inject, Param, Post } from "@nestjs/common";
import {
  AuthzContext,
  AUTHZ_SERVICE_IDENTITY,
  attachOutboundPropagation,
} from "authz-nestjs";
import type { RequestContext } from "authz-nestjs";
import axios, { AxiosInstance } from "axios";

const DOWNSTREAM = process.env.DOWNSTREAM_URL || "http://localhost:5002";

/**
 * Business routes — no authorization annotations. `@AuthzContext()` exposes the
 * validated RequestContext built by the global guard. Outbound forwarding uses an
 * axios instance with the library's propagation interceptor: it auto-attaches the
 * user JWT, this service's token, and the correlation/request ids for any call
 * made inside an authorized request (the inbound context flows via AsyncLocalStorage).
 */
@Controller("orders")
export class OrdersController {
  private readonly downstream: AxiosInstance;

  constructor(@Inject(AUTHZ_SERVICE_IDENTITY) serviceIdentity: any) {
    this.downstream = axios.create({ timeout: 5000 });
    attachOutboundPropagation(this.downstream, { serviceIdentity: serviceIdentity ?? undefined });
  }

  @Get(":id")
  get(@Param("id") id: string, @AuthzContext() ctx: RequestContext) {
    return { id, by: ctx.userId };
  }

  @Get(":id/audit")
  audit(@Param("id") id: string) {
    return { id, audit: true };
  }

  @Post()
  create() {
    return { created: true };
  }

  @Post(":id")
  update(@Param("id") id: string, @AuthzContext() ctx: RequestContext) {
    return {
      id,
      updated: true,
      seenServiceName: ctx?.serviceName || "none",
      seenAuthType: ctx?.authenticationType || "none",
      seenCorrelationId: ctx?.correlationId || "none",
      seenUserId: ctx?.userId || "none",
    };
  }

  @Post(":id/forward")
  async forward(@Param("id") id: string, @Body() body: any) {
    try {
      const r = await this.downstream.post(`${DOWNSTREAM}/orders/${id}`, body ?? {});
      return {
        forwarded: true,
        downstreamStatus: r.status,
        sentHeaders: {
          hasServiceToken: !!(
            r.config?.headers?.["X-Service-Token"] || r.config?.headers?.["x-service-token"]
          ),
          hasAuthorization: !!(
            r.config?.headers?.["Authorization"] || r.config?.headers?.["authorization"]
          ),
          correlationId:
            r.config?.headers?.["X-Correlation-Id"] || r.config?.headers?.["x-correlation-id"],
          requestId: r.config?.headers?.["X-Request-Id"] || r.config?.headers?.["x-request-id"],
        },
        downstream: r.data,
      };
    } catch (e: any) {
      if (e.response) {
        return {
          forwarded: false,
          downstreamStatus: e.response.status,
          downstream: e.response.data,
        };
      }
      throw e;
    }
  }
}
